pageserver/src/tenant/timeline/compaction.rs

```diff
@@ -454,6 +455,14 @@ impl Timeline {
             }
             deltas_to_compact.push(l.download_and_keep_resident().await?);
             prev_lsn_end = lsn_range.end;
+
+            if deltas_to_compact.len() >= L0_COMPACTION_MAX_NUM_DELTA_LAYERS {
```

Comment:
There's an assumption here that when we have more layers than normal, they're all going to be big.  I guess that's usually accurate, but there could be some edge cases: like if a  pageserver restarted repeatedly or a tenant was migrated repeatedly, generating many smaller L0s.

So perhaps we should accumulate the size of layers in `deltas_to_compact`, and then apply the condition that it should not exceed `compaction_threshold * checkpoint_distance`, so that have a max dataset approximately the size of one "normal" ten-layer compaction of large layers.